File: detox/detox/ios/Detox/Utilities/NSThread+DetoxUtils.m
//
//  NSThread+DetoxUtils.m
//  Detox
//
//  Created by Leo Natan (Wix) on 7/16/20.
//  Copyright © 2020 Wix. All rights reserved.
//

#import "NSThread+DetoxUtils.h"
#import "DTXAddressInfo.h"

@implementation NSThread (DetoxUtils)

+ (NSString*)dtx_demangledCallStackSymbols
{
	return [self dtx_demangledCallStackSymbolsForReturnAddresses:self.callStackReturnAddresses startIndex:1];
}

+ (NSString*)dtx_demangledCallStackSymbolsForReturnAddresses:(NSArray<NSNumber*>*)returnAddresses startIndex:(NSInteger)startIndex
{
	NSArray* symbols = [returnAddresses dtx_mapObjectsUsingBlock:^id _Nonnull(NSNumber * _Nonnull obj, NSUInteger idx) {
		return [[[DTXAddressInfo alloc] initWithAddress:obj.unsignedIntegerValue] formattedDescriptionForIndex:idx];
	}];
	return [NSString stringWithFormat:@"(\n\t%@\n)", [symbols componentsJoinedByString:@"\n\t"]];
}

@end

File: detox/ios/Detox/Next/Utilities/NSThread+DetoxUtils.swift
//
//  NSThread+DetoxUtils.swift
//  Detox
//
//  Created by Leo Natan (Wix) on 5/31/20.
//  Copyright © 2020 Wix. All rights reserved.
//

import Foundation

@objc
extension Thread {
	@objc(dtx_demangledCallStackSymbols)
	open class var demangledCallStackSymbols : String {
		return demangledCallStackSymbolsFor(returnAddresses: callStackReturnAddresses, startIndex: 2)
	}
	
	internal static func demangledCallStackSymbolsFor(returnAddresses: [NSNumber], startIndex: Int) -> String {
		let symbols = returnAddresses[startIndex..<returnAddresses.count].enumerated().map {
			DTXAddressInfo(address: UInt(truncating: $1)).formattedDescription(for: UInt($0))
		}
		return "(\n\t\(symbols.joined(separator: "\n\t"))\n)"
	}
}
